File: packages/go/preflight.go
package rdk

import (
	"context"
	"fmt"
	"math/big"
)

// CheckStatus is the status of a single preflight check.
type CheckStatus string

const (
	// CheckOK indicates the check passed.
	CheckOK CheckStatus = "ok"
	// CheckWarn indicates a non-fatal warning.
	CheckWarn CheckStatus = "warn"
	// CheckFail indicates a hard failure.
	CheckFail CheckStatus = "fail"
)

// PreflightCheck is a single readiness check result.
type PreflightCheck struct {
	ID     string
	Label  string
	Status CheckStatus
	Detail string
	Hint   string
}

// PreflightResult is the aggregate result of the preflight checks.
type PreflightResult struct {
	// OK is true when no check failed (warnings are allowed).
	OK     bool
	Checks []PreflightCheck
}

// PreflightOptions configures the preflight checks.
type PreflightOptions struct {
	// Config is a rollup config to validate. Nil skips the config check.
	Config *RollupConfig
	// SignerAddress is the operator/signer address, to check balance/presence.
	SignerAddress string
	// ExpectedNetwork asserts the client is pointed at this network.
	ExpectedNetwork string
	// FeeBufferUqor is extra uqor required on top of the stake. Empty defaults to
	// 1 QOR (1000000 uqor).
	FeeBufferUqor string
}

// CheckPreflight runs the preflight checks against a client.
func CheckPreflight(ctx context.Context, client *RdkClient, options PreflightOptions) PreflightResult {
	checks := []PreflightCheck{}

	params, err := client.Params(ctx)
	haveParams := err == nil
	if haveParams {
		checks = append(checks, PreflightCheck{
			ID: "rest", Label: "REST endpoint reachable", Status: CheckOK,
			Detail: client.Network.Endpoints.Rest,
		})
		minStakeQor, _ := UqorToQor(params.MinStakeForRollup, 0)
		burnPct := new(big.Float).SetFloat64(0)
		if r, ok := new(big.Float).SetString(params.RollupCreationBurnRate); ok {
			burnPct = r.Mul(r, big.NewFloat(100))
		}
		checks = append(checks, PreflightCheck{
			ID: "params", Label: "Module parameters readable", Status: CheckOK,
			Detail: fmt.Sprintf("min stake %s QOR, burn %s%%", minStakeQor, burnPct.Text('g', -1)),
		})
	} else {
		checks = append(checks, PreflightCheck{
			ID: "rest", Label: "REST endpoint reachable", Status: CheckFail,
			Detail: err.Error(),
			Hint:   "Set QORE_REST_URL to a reachable node REST (LCD) endpoint.",
		})
	}

	if options.ExpectedNetwork != "" {
		match := client.Network.Name == options.ExpectedNetwork
		status := CheckOK
		hint := ""
		if !match {
			status = CheckWarn
			hint = "Expected " + options.ExpectedNetwork + "."
		}
		checks = append(checks, PreflightCheck{
			ID: "network", Label: "Network matches expectation", Status: status,
			Detail: fmt.Sprintf("client is %s (%s)", client.Network.Name, client.Network.ChainID),
			Hint:   hint,
		})
	}

	if options.Config != nil {
		r := ValidateRollupConfig(*options.Config)
		switch {
		case !r.Valid:
			checks = append(checks, PreflightCheck{
				ID: "config", Label: "Rollup config valid", Status: CheckFail,
				Detail: r.Errors[0], Hint: "Fix the configuration errors before creating.",
			})
		case len(r.Warnings) > 0:
			checks = append(checks, PreflightCheck{
				ID: "config", Label: "Rollup config valid", Status: CheckWarn, Detail: r.Warnings[0],
			})
		default:
			checks = append(checks, PreflightCheck{
				ID: "config", Label: "Rollup config valid", Status: CheckOK,
				Detail: "compatibility matrix satisfied",
			})
		}
	}

	if options.SignerAddress != "" {
		checks = append(checks, PreflightCheck{
			ID: "signer", Label: "Signer configured", Status: CheckOK, Detail: options.SignerAddress,
		})
		if haveParams {
			bal, balErr := client.Rest.GetBalance(ctx, options.SignerAddress, "")
			if balErr != nil {
				checks = append(checks, PreflightCheck{
					ID: "balance", Label: "Balance readable", Status: CheckWarn, Detail: balErr.Error(),
				})
			} else {
				stake, _ := new(big.Int).SetString(params.MinStakeForRollup, 10)
				if stake == nil {
					stake = big.NewInt(0)
				}
				bufStr := options.FeeBufferUqor
				if bufStr == "" {
					bufStr = "1000000"
				}
				buffer, _ := new(big.Int).SetString(bufStr, 10)
				if buffer == nil {
					buffer = big.NewInt(0)
				}
				required := new(big.Int).Add(stake, buffer)
				have, _ := new(big.Int).SetString(bal, 10)
				if have == nil {
					have = big.NewInt(0)
				}
				ok := have.Cmp(required) >= 0
				haveQor, _ := UqorToQor(have.String(), 0)
				reqQor, _ := UqorToQor(required.String(), 0)
				status := CheckOK
				hint := ""
				if !ok {
					status = CheckFail
					hint = "Fund the operator account (see the keys & funding guide)."
				}
				checks = append(checks, PreflightCheck{
					ID: "balance", Label: "Balance covers stake + fees", Status: status,
					Detail: fmt.Sprintf("have %s QOR, need ~%s QOR", haveQor, reqQor), Hint: hint,
				})
			}
		}
	} else {
		checks = append(checks, PreflightCheck{
			ID: "signer", Label: "Signer configured", Status: CheckWarn, Detail: "no signer",
			Hint: "Set QORE_OPERATOR_PRIVATE_KEY_HEX or QORE_MNEMONIC to create/operate.",
		})
	}

	ok := true
	for _, c := range checks {
		if c.Status == CheckFail {
			ok = false
		}
	}
	return PreflightResult{OK: ok, Checks: checks}
}

File: packages/rust/src/preflight.rs
//! Preflight checks -- the engine behind a `doctor` command. Verifies, in plain
//! language, that a developer is ready to create/operate a rollup: endpoints
//! reachable, network as expected, module params readable, config valid, a
//! signer configured, and the operator balance covering the stake plus a fee
//! buffer.

use crate::client::RdkClient;
use crate::config::{validate_rollup_config, RollupConfig};
use crate::constants::{Network, BASE_DENOM};
use crate::utils::denom::uqor_to_qor_default;

/// A check outcome.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CheckStatus {
    /// The check passed.
    Ok,
    /// The check passed with a caveat.
    Warn,
    /// The check failed.
    Fail,
}

/// A single preflight check result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PreflightCheck {
    /// A stable identifier.
    pub id: String,
    /// A human-readable label.
    pub label: String,
    /// The outcome.
    pub status: CheckStatus,
    /// An optional detail line.
    pub detail: Option<String>,
    /// An optional remediation hint.
    pub hint: Option<String>,
}

/// The overall preflight result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PreflightResult {
    /// True when no check failed (warnings are allowed).
    pub ok: bool,
    /// The individual checks.
    pub checks: Vec<PreflightCheck>,
}

/// Options for [`check_preflight`].
#[derive(Default)]
pub struct PreflightOptions<'a> {
    /// A rollup config to validate.
    pub config: Option<&'a RollupConfig>,
    /// The operator/signer address.
    pub signer_address: Option<String>,
    /// Assert the client is pointed at this network.
    pub expected_network: Option<Network>,
    /// Extra uqor buffer to require on top of the stake (fees). Default 1 QOR.
    pub fee_buffer_uqor: Option<u128>,
}

/// Run the preflight checks against a client.
pub fn check_preflight(client: &RdkClient, options: PreflightOptions<'_>) -> PreflightResult {
    let mut checks: Vec<PreflightCheck> = Vec::new();
    let mut params = None;

    match client.params() {
        Ok(p) => {
            checks.push(PreflightCheck {
                id: "rest".to_string(),
                label: "REST endpoint reachable".to_string(),
                status: CheckStatus::Ok,
                detail: Some(client.network.endpoints.rest.clone()),
                hint: None,
            });
            let stake_qor = uqor_to_qor_default(&p.min_stake_for_rollup)
                .unwrap_or_else(|_| p.min_stake_for_rollup.clone());
            let burn_pct = p
                .rollup_creation_burn_rate
                .parse::<f64>()
                .map(|r| r * 100.0)
                .unwrap_or(0.0);
            checks.push(PreflightCheck {
                id: "params".to_string(),
                label: "Module parameters readable".to_string(),
                status: CheckStatus::Ok,
                detail: Some(format!("min stake {stake_qor} QOR, burn {burn_pct}%")),
                hint: None,
            });
            params = Some(p);
        }
        Err(e) => {
            checks.push(PreflightCheck {
                id: "rest".to_string(),
                label: "REST endpoint reachable".to_string(),
                status: CheckStatus::Fail,
                detail: Some(e.to_string()),
                hint: Some("Set the REST (LCD) endpoint to a reachable node.".to_string()),
            });
        }
    }

    if let Some(expected) = options.expected_network {
        let matches = client.network.name == expected;
        checks.push(PreflightCheck {
            id: "network".to_string(),
            label: "Network matches expectation".to_string(),
            status: if matches {
                CheckStatus::Ok
            } else {
                CheckStatus::Warn
            },
            detail: Some(format!(
                "client is {} ({})",
                client.network.name.as_str(),
                client.network.chain_id
            )),
            hint: if matches {
                None
            } else {
                Some(format!("Expected {}.", expected.as_str()))
            },
        });
    }

    if let Some(config) = options.config {
        let r = validate_rollup_config(config);
        let status = if r.valid {
            if r.warnings.is_empty() {
                CheckStatus::Ok
            } else {
                CheckStatus::Warn
            }
        } else {
            CheckStatus::Fail
        };
        let detail = if r.valid {
            r.warnings
                .first()
                .cloned()
                .or_else(|| Some("compatibility matrix satisfied".to_string()))
        } else {
            r.errors.first().cloned()
        };
        checks.push(PreflightCheck {
            id: "config".to_string(),
            label: "Rollup config valid".to_string(),
            status,
            detail,
            hint: if r.valid {
                None
            } else {
                Some("Fix the configuration errors before creating.".to_string())
            },
        });
    }

    if let Some(addr) = &options.signer_address {
        checks.push(PreflightCheck {
            id: "signer".to_string(),
            label: "Signer configured".to_string(),
            status: CheckStatus::Ok,
            detail: Some(addr.clone()),
            hint: None,
        });
        if let Some(p) = &params {
            match client.rest.get_balance(addr, BASE_DENOM) {
                Ok(bal) => {
                    let stake = p.min_stake_for_rollup.parse::<u128>().unwrap_or(0);
                    let buffer = options.fee_buffer_uqor.unwrap_or(1_000_000);
                    let required = stake + buffer;
                    let have = bal.parse::<u128>().unwrap_or(0);
                    let ok = have >= required;
                    checks.push(PreflightCheck {
                        id: "balance".to_string(),
                        label: "Balance covers stake + fees".to_string(),
                        status: if ok {
                            CheckStatus::Ok
                        } else {
                            CheckStatus::Fail
                        },
                        detail: Some(format!(
                            "have {} QOR, need ~{} QOR",
                            uqor_to_qor_default(&bal).unwrap_or(bal.clone()),
                            uqor_to_qor_default(&required.to_string())
                                .unwrap_or_else(|_| required.to_string())
                        )),
                        hint: if ok {
                            None
                        } else {
                            Some("Fund the operator account.".to_string())
                        },
                    });
                }
                Err(e) => {
                    checks.push(PreflightCheck {
                        id: "balance".to_string(),
                        label: "Balance readable".to_string(),
                        status: CheckStatus::Warn,
                        detail: Some(e.to_string()),
                        hint: None,
                    });
                }
            }
        }
    } else {
        checks.push(PreflightCheck {
            id: "signer".to_string(),
            label: "Signer configured".to_string(),
            status: CheckStatus::Warn,
            detail: Some("no signer".to_string()),
            hint: Some("Provide a signer to create/operate.".to_string()),
        });
    }

    let ok = checks.iter().all(|c| c.status != CheckStatus::Fail);
    PreflightResult { ok, checks }
}
